packages/lib/src/vis/shared/Html.tsx

```diff
@@ -1,48 +1,42 @@
+import { assertNonNull } from '@h5web/shared';
 import { useThree } from '@react-three/fiber';
-import type { ReactNode } from 'react';
+import type { PropsWithChildren } from 'react';
 import { useLayoutEffect, useState } from 'react';
-import ReactDOM from 'react-dom';
+import ReactDOM, { createPortal } from 'react-dom';
 
 interface Props {
-  overflowCanvas?: boolean; // allow children to overflow above axes
-  container?: HTMLElement;
-  children?: ReactNode;
+  overflowCanvas?: boolean;
 }
 
-function Html(props: Props) {
-  const {
-    overflowCanvas = false,
-    container: customContainer,
-    children,
-  } = props;
+function Html(props: PropsWithChildren<Props>) {
+  const { overflowCanvas = false, children } = props;
 
   const r3fRoot = useThree((state) => state.gl.domElement.parentElement);
-  const canvasWrapper = r3fRoot?.parentElement;
+  assertNonNull(r3fRoot);
 
-  // Choose DOM container to which to append `renderTarget`
-  // (with `canvasWrapper`, `Html` children are allowed to overflow above the axes)
-  const container =
-    customContainer || (overflowCanvas ? canvasWrapper : r3fRoot);
+  const canvasWrapper = r3fRoot.parentElement;
+  assertNonNull(canvasWrapper);
 
-  const [renderTarget] = useState(() => document.createElement('div'));
+  const portalContainer = overflowCanvas ? canvasWrapper : r3fRoot;
 
-  useLayoutEffect(() => {
-    ReactDOM.render(<>{children}</>, renderTarget); // eslint-disable-line react/jsx-no-useless-fragment
-  }, [children, renderTarget]);
+  const [renderContainer] = useState(() => {
+    const div = document.createElement('div');
+    div.setAttribute('hidden', '');
+    return div;
+  });
 
   useLayoutEffect(() => {
-    return () => {
-      ReactDOM.unmountComponentAtNode(renderTarget);
```

Comment:
Turns out that having this in a separate effect with only `renderTarget` as dep was a bug. `renderTarget` never changes—it's `container` that changes—so the first clean-up function was not always called when it should have been.

The bug is gone now, since we always append `renderTarget` (now `renderContainer`) into `r3fRoot`, which never changes (so both clean-up functions would get called when `Html` unmounts), but it's still cleaner to have them in a single clean-up function.